src/main/java/rx/internal/operators/OperatorPublish.java

```diff
@@ -15,367 +15,736 @@
             /*
-             * when receiving a request from a subscriber we reset 'outstanding' to the lowest of all subscribers
+             * In case an exception is thrown in the loop, we need to set emitting back to false
+             * on the way out (the exception will propagate up) so if it bounces back and
+             * onError is called, its dispatch() call will have the opportunity to emit it.
+             * However, if we want to exit regularly, we will set the emitting to false (+ other operations)
+             * atomically so we want to prevent the finally part to accidentally unlock some other
+             * emissions happening between the two synchronized blocks.
              */
-            outstandingRequests = lowest;
-            return lowest;
+            boolean skipFinal = false;
+            try {
+                for (;;) {
+                    /*
+                     * See if the queue is empty; since we need this information multiple
+                     * times later on, we read it one.
+                     * Although the queue can become non-empty in the mean time, we will
+                     * detect it through the missing flag and will do another iteration.
+                     */
+                    boolean empty = queue.isEmpty();
+                    // if the queue is empty and the terminal event was received, quit
+                    // and don't bother restoring emitting to false: no further activity is
+                    // possible at this point
+                    if (checkTerminated(terminalEvent, empty)) {
+                        skipFinal = true;
+                        return;
+                    }
+                    
+                    // We have elements queued. Note that due to the serialization nature of dispatch()
+                    // this loop is the only one which can turn a non-empty queue into an empty one
+                    // and as such, no need to ask the queue itself again for that.
+                    if (!empty) {
+                        // We take a snapshot of the current child-subscribers.
+                        // Concurrent subscribers may miss this iteration, but it is to be expected
+                        @SuppressWarnings("unchecked")
+                        InnerProducer<T>[] ps = producers.get();
+                        
+                        int len = ps.length;
+                        // Let's assume everyone requested the maximum value.
+                        long maxRequested = Long.MAX_VALUE;
+                        // count how many have triggered unsubscription
+                        int unsubscribed = 0;
+                    
+                        // Now find the minimum amount each child-subscriber requested
+                        // since we can only emit that much to all of them without violating
+                        // backpressure constraints
+                        for (InnerProducer<T> ip : ps) {
```

Comment:
You can only request the common minimum amount from upstream if the front-queue is bounded, even if some child subscribers request more. These may run ahead but then stay idle until the 'slower ones' catch up.